sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -137,8 +138,16 @@ case class EnsureRequirements(
         Some(finalCandidateSpecs.values.maxBy(_.numPartitions))
       }
 
+      // Check if 1) all children are of `KeyGroupedPartitioning` and 2) they are all compatible
+      // with each other. If both are true, skip shuffle.
+      val allCompatible = childrenIndexes.sliding(2).map {
```

Comment:
In the SPIP, it is mentioned that if there are few large partitions, we will still do a shuffle. We are not trying to do that here?

Reply:
This is handled at the moment via config `spark.sql.shuffle.partitions`: in `EnsureRequirements` if the number of partitions is smaller than this number, Spark will insert shuffle.